internal/rbd/controllerserver.go

```diff
@@ -332,6 +310,36 @@ func flattenParentImage(ctx context.Context, rbdVol *rbdVolume, cr *util.Credent
 	return nil
 }
 
+// repairExistingVolume checks the existing volume or snapshot and makes sure
+// that the state is corrected to what was requested. It is needed to call this
+// when the process of creating a volume was interrupted.
+func (cs *ControllerServer) repairExistingVolume(ctx context.Context, req *csi.CreateVolumeRequest,
+	cr *util.Credentials, rbdVol *rbdVolume, rbdSnap *rbdSnapshot) (*csi.CreateVolumeResponse, error) {
+	if rbdSnap != nil {
+		// check if image depth is reached limit and requires flatten
+		err := checkFlatten(ctx, rbdVol, cr)
+		if err != nil {
+			return nil, err
+		}
+
+		err = rbdSnap.repairEncryptionConfig(&rbdVol.rbdImage)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		// continue/restart allocating the volume in case it
+		// should be thick-provisioned
+		if cs.isThickProvisionRequest(req) {
```

Comment:
can we have a check for repairing only fresh volume, not a PVC clone?

Reply:
I do not think that is needed? Creating from a snapshot will have `rbdSnap` above set, and a clone will already have the thick-provisioned metadata in the image.

We could check if the VolumeContentSource is empty, but would like to understand what corner-case I'm missing.

Reply:
Can you please check it. the image metadata will get copied to the cloned images? just now tested it I didn't see image metadata getting copied to clone, Am I missing something here?

```
[root@dhcp53-170 rbd]# kubectl get pv pvc-9533729e-6947-4612-8fb7-58a8b01985fa -oyaml |grep -i imageName
      imageName: csi-vol-57e4646e-be17-11eb-9d80-0242ac110005
[root@dhcp53-170 rbd]# 
[root@dhcp53-170 rbd]# kubectl get pv pvc-efd86ad4-3fb9-40bb-b1f0-91b8cc4974d1 -oyaml |grep -i imageName
      imageName: csi-vol-665bd1ae-be17-11eb-9d80-0242ac110005


sh-4.4# rbd image-meta ls replicapool/csi-vol-665bd1ae-be17-11eb-9d80-0242ac110005    
There are 0 metadata on this image.
sh-4.4# rbd image-meta ls replicapool/csi-vol-57e4646e-be17-11eb-9d80-0242ac110005
There is 1 metadatum on this image:

Key                                  Value
.rbd.csi.ceph.com/thick-provisioned  true 
sh-4.4# rbd image-meta ls replicapool/csi-snap-60cc6470-be17-11eb-9d80-0242ac110005
There are 0 metadata on this image.
sh-4.4# 
```

Currently, I don't think of a corner case but I thought it's good to have an empty Content check.

Reply:
hmm, I though this was well tested, also in combination with encryption... will try it out again

Reply:
Indeed, I probably misremember things. A clone (or even deep-copy) does not take the metadata from the source to the destination.

That makes this approach incomplete....

Reply:
I guess improvements to the cloning/snapshotting will be handled through #2077. This PR only addresses the restarting of CreateVolume when a thick-provisioning is in progress.

Reply:
@Madhu-1: now only repairing the thick-provisioning for a new (non clone/restore) volume,